workflow/scripts/general_html_report.py

```diff
@@ -0,0 +1,268 @@
+from jinja2 import Template
+import json
+from jsonschema import validate
+from jsonschema.exceptions import ValidationError
+import re
+import sys
+import time
+import yaml
+import pandas as pd
+import numpy as np
+
+
+def validate_dict(d: dict, schema_path: str):
+    with open(schema_path, "r") as f:
+        try:
+            validate(instance=d, schema=yaml.safe_load(f))
+        except ValidationError as ve:
+            print(f"error: failed to validate general report config:", file=sys.stderr)
+            print(ve, file=sys.stderr)
+            sys.exit(1)
+
+
+def validate_table_data(table: list) -> bool:
+    if len(table) == 0:
+        raise ValueError("empty table")
+
+    n_cols = len(table[0])
+    cols = table[0].keys()
+
+    for i, row in enumerate(table, start=1):
+        if len(row) != n_cols:
+            raise ValueError(f"expected {n_cols} columns in row {i}, found {len(row)}")
+        if set(row.keys()) != set(cols):
+            raise ValueError(
+                f"expected columns {', '.join(repr(x) for x in cols)} "
+                f"in row {i}, found {', '.join(repr(x) for x in row.keys())}"
+            )
+    return True
+
+
+def fix_relative_uri(uri: str, depth: int) -> str:
+    if re.match("^([^/]+:|/)", uri):
+        return uri
+    return depth * "../" + uri
+
+
+def parse_multiqc_config(d: dict):
+    with open(d) as f:
+        test = yaml.safe_load(f)
+        if test["table_columns_placement"]:
+            for k in test["table_columns_placement"]:
+                to_keep = list(test["table_columns_placement"][k].keys())
+
+        if test["custom_data"]:
+            for k in test["custom_data"]:
+                if "generalstats" in test["custom_data"][k]["plot_type"]:
+                    for h in test["custom_data"][k]["pconfig"]:
+                        for g in h.keys():
+                            to_keep.append(g)
+    return to_keep
```

Comment:
I'm not really sure what this function should do, but there are some issues with it. `d` shouldn't be a `dict`, but a path (either `pathlib.Path` or `str`). The variable `to_keep` will be overwritten in every iteration inside the for loop in the first if statement (which will fail with a KeyError if the key is missing), and the variable might even be undefined by the end of the function since it is defined within the first if statement. This is also a reason why I think a test case for MultiQC should be added.

Reply:
Also, there's no need to nest the if statements in the with statement. They can go outside.

Reply:
I've added a multiqc_data test file now so you can test it. The function removes columns in the multiqc general stats according to what has been configured in the multiqc_config

Reply:
Great! I'll give it a try this afternoon!

Reply:
The integration test works for me too. Nice!